<h1>Project: Data Analytics Project</h1>

<h2>United Kingdom Road Accident Data Analysis</h2>
<p>Data Analyst: Steven R. Maraig</p>

<h1>Importing Libraries</h1>

In [1]:
import seaborn as sns
import pandas as pd 
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import f_oneway

<h1>Creating a Data Frame</h1>

In [2]:
accident = pd.read_csv('dataset//accident_data.csv')


In [3]:
accident.shape

(660679, 14)

In [4]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1>Checking Null values</h1>

In [5]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1>Handle missing values</h1>

In [6]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna('Unknown Road Type')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('Unknown Weather Condition')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])

accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Changing data type</h1>

In [7]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [8]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')
accident['Accident Date'] = accident['Accident Date'].astype('category')



In [9]:
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident['Accident Date'] = accident['Accident Date'].str.replace('/','-')

In [10]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'],dayfirst=True, errors='coerce')
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h2>Extracting date information using pandas date time </h2>

In [11]:
accident['Year'] = accident['Accident Date'].dt.year
accident['month'] = accident['Accident Date'].dt.month
accident['day_of_week'] = accident['Accident Date'].dt.day_name()

<h4>Changing the data type of the new column (nakalimutan ko po ito hehe)</h4>

In [12]:
accident['Year'] = accident['Year'].astype('category')
accident['month'] = accident['month'].astype('category')
accident['day_of_week'] = accident['Year'].astype('category')

In [13]:
accident['month']

0         6
1         7
2         8
3         8
4         9
         ..
660674    2
660675    2
660676    2
660677    2
660678    2
Name: month, Length: 660679, dtype: category
Categories (12, int32): [1, 2, 3, 4, ..., 9, 10, 11, 12]

<h2>How many incident happend under ligh condition during 2022</h2>

<h1>Insight No1. </h1>
<h3>74.64% of the accident occur during 2022 is Daylight conditions, Hypothesis: There are high percentage of accident during daylight conditions</h3>

In [14]:
during_2019 = accident[accident['Year'] == 2019]
during_2022 = accident[accident['Year'] == 2022]
during_2020 = accident[accident['Year'] == 2020]
during_2021 = accident[accident['Year'] == 2021]
during_2019['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

<h3>What is the average number of casualties over the following years?</h3>

<h1>Insight No2. </h1>
<h3>Over the period from 2019 to 2022, the data reveals that the average number of casualties per year was approximately 1.3.</h3>

In [15]:
casuality_per_year = accident.groupby(['Year'])['Number_of_Casualties'].mean()
casuality_per_year



Year
2019    1.360569
2020    1.353559
2021    1.358243
2022    1.355341
Name: Number_of_Casualties, dtype: float64

<h3>What is the distribution of accident severity?</h3>


<h1>Insight No3. </h1>
<h3>Vast majority of accidents are 85.34% 'Slight,' There are significantly fewer Serious with 13.35% , and even fewer with Fatal 1.31% accidents, with 8,661 occurrences.</h3>

In [16]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h3>During 2020 which  areas are most prone to accidents?</h3>


<h1>Insight No4. </h1>
<h3>In rural areas, most accidents are slight, with 50,722 slight accidents compared to 9,774 serious and 1,495 fatal accidents. Urban areas also have many slight accidents (94,402) but with 13,346 serious and 846 fatal accidents</h3>

In [17]:
fatality_area = during_2020.groupby(['Urban_or_Rural_Area','Accident_Severity']).size()
fatality_area

Urban_or_Rural_Area  Accident_Severity
Rural                Fatal                 1495
                     Serious               9774
                     Slight               50722
Unallocated          Fatal                    0
                     Serious                  1
                     Slight                   5
Urban                Fatal                  846
                     Serious              13346
                     Slight               94402
dtype: int64

<h2>Do vehicle type actually affect the severity of the accident?</h2>


<h1>Insight No5. </h1>
<h3>During 2020 Cars account for 75.7% of all fatalities (1,772 out of 2,341), making them the most accident-prone vehicle type far exceeding other vehicles like buses or coaches (3.9%) and motorcycles over 500cc (4.2%). This highlights the urgent need for targeted safety measures to address car-related risks.</h3>

In [18]:
vehicle_severity = during_2020.groupby(['Vehicle_Type','Accident_Severity']).size()
vehicle_severity.unstack()

Accident_Severity                      Fatal  Serious  Slight
Vehicle_Type                                                 
Agricultural vehicle                       7       74     481
Bus or coach (17 or more pass seats)      91      841    5777
Car                                     1772    17384  108738
Data missing or out of range               0        0       6
Goods 7.5 tonnes mgw and over             52      622    3854
Goods over 3.5t. and under 7.5t           15      217    1313
Minibus (8 - 16 passenger seats)           9       72     449
Motorcycle 125cc and under                41      560    3503
Motorcycle 50cc and under                 24      243    1497
Motorcycle over 125cc and up to 500cc     22      258    1645
Motorcycle over 500cc                     98      908    5739
Other vehicle                             23      217    1293
Pedal cycle                                2       17      44
Ridden horse                               0        0       0
Taxi/Private hire car                     48      490    2983
Van / Goods 3.5 tonnes mgw or under      137     1218    7807

<h2>Now what is the common road surface condition in vehicle type car during 2020?</h2>


<h1>Insight No6. </h1>
<h3>In 2020, the majority of "car" type accidents occurred on dry road surfaces, accounting for 66.67% (85,266 incidents) of all cases far exceeding wet or damp conditions (30.49%), frost or ice (2.28%), snow (0.35%), and flooding (0.19%).</h3>

In [19]:
car_2020 = during_2020[during_2020['Vehicle_Type'] == 'Car']
car_2020['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     85266
Wet or damp             39019
Frost or ice             2915
Snow                      447
Flood over 3cm. deep      247
Name: count, dtype: int64

<h2>Now what is the common Weather condition in vehicle type car during 2020?</h2>


<h1>Insight No7. </h1>
<h3>77.78% of all cases (99,486 incidents) occurred during fine weather conditions with no high winds, far exceeding other conditions such as raining no high winds (12.82%) and fine + high winds (1.67%). This highlights that favorable weather does not necessarily equate to safer driving conditions.</h3>

In [20]:
car_2020['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds           99486
Raining no high winds        16393
Other                         3177
Unknown Weather Condition     2938
Raining + high winds          2367
Fine + high winds             2134
Fog or mist                    716
Snowing no high winds          538
Snowing + high winds           145
Name: count, dtype: int64

<h2>what is the common District an accident will occur in vehicle type car during 2020?</h2>


<h1>Insight No.8 </h1>
<h3>The majority of accidents occurred in the Birmingham district, accounting for 2,048 incidents—the highest among all districts—followed by Leeds (1,794) and Manchester (1,370). </h3>

In [21]:
car_2020['District Area'].value_counts()


District Area
Birmingham                   2048
Leeds                        1794
Manchester                   1370
Glasgow City                 1312
Bradford                     1205
                             ... 
Cheshire West and Chester       0
Cheshire East                   0
County Durham                   0
Central Bedfordshire            0
Northumberland                  0
Name: count, Length: 422, dtype: int64

In [22]:
accident.columns

Index(['Index', 'Accident_Severity', 'Accident Date', 'Latitude',
       'Light_Conditions', 'District Area', 'Longitude',
       'Number_of_Casualties', 'Number_of_Vehicles', 'Road_Surface_Conditions',
       'Road_Type', 'Urban_or_Rural_Area', 'Weather_Conditions',
       'Vehicle_Type', 'Year', 'month', 'day_of_week'],
      dtype='object')

<h2>what is the accident severity trends over the past few years do the fatality increases?</h2>


<h1>Insight No.9 </h1>
<h3>The data shows a consistent decline in fatal accidents from 2019 to 2022, with a total reduction of 42.9% over the four years. Specifically, fatalities dropped by 13.74% from 2019 (2,714 accidents) to 2020 (2,341), by 12.13% from 2020 to 2021 (2,057), and by a significant 24.69% from 2021 to 2022 (1,549).</h3>

In [23]:
#Before po ma realize na naka int po data type ni year
#gawa po ito ni claude
# year_dfs = [during_2019, during_2020, during_2021, during_2022]
# years = list(range(2019, 2023))


# fatal_accidents = {}


# for year, df in zip(years, year_dfs):
 
#     fatal_accidents[year] = df[df['Accident_Severity'] == 'Fatal']
 
#     print(f"Fatal accidents in {year}: {fatal_accidents[year].shape}")

In [24]:
years = accident.groupby(['Year','Accident_Severity']).size()
years

Year  Accident_Severity
2019  Fatal                  2714
      Serious               24322
      Slight               155079
2020  Fatal                  2341
      Serious               23121
      Slight               145129
2021  Fatal                  2057
      Serious               21997
      Slight               139500
2022  Fatal                  1549
      Serious               18777
      Slight               124093
dtype: int64

<h2>Which months have the highest/lowest number of accidents?</h2>


<h1>Insight No.10 </h1>
<h3>November recorded the highest number of accidents at 60,424, accounting for approximately 9.29% of the total annual accidents, while February had the lowest with 49,491 incidents, contributing 7.61%. Interestingly, the data shows that accident rates tend to peak during late autumn (October–November), possibly due to reduced daylight hours and worsening weather conditions.</h3>

In [25]:
accident['month'].value_counts()

month
11    60424
10    59580
7     57445
6     56481
9     56455
5     56352
3     54086
8     53913
1     52872
12    51836
4     51744
2     49491
Name: count, dtype: int64

<h2>Since the answer in insight number 10, that the peak of accident happening is in octuber and november which is rainy season the question is if in this certain months do weather conditions affect the accident occuring?</h2>


<h1>Insight No.11 </h1>
<h3>No, most accidents occurring during October and November happened under 'fine no high winds' weather conditions, accounting for 80.26% (47,822 incidents), while 'raining no high winds' accounted for only 12.78% (7,615 incidents). </h3>

In [26]:
# october_november = accident[(accident['month'] == '10') & (accident['month'] == '11')]
# (accident['month'] == '11').isnull().sum()

# Example with AND logic for different columns
# result = accident[(accident['month'] == 10) & (accident['month'] == 11) ] 
# result

october_november = accident[accident['month'] == 10 & 11]
season_server_weather = october_november.groupby(['Weather_Conditions','Accident_Severity']).size()
october_november['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds           47822
Raining no high winds         7615
Unknown Weather Condition     1310
Other                         1188
Raining + high winds           710
Fine + high winds              579
Fog or mist                    278
Snowing no high winds           67
Snowing + high winds            11
Name: count, dtype: int64

<h2>What month have the highest fatality rate in 2022>?</h2>


<h1>Insight No.12 </h1>
    <h3>The highest fatality rate occurs in August, with 160 fatalities accounting for 1.32% of total accidents—making it the month with the highest percentage of fatalities relative to total accidents.</h3>

In [27]:

during_2022.groupby(['month','Accident_Severity']).size().unstack().T

month                 1     2      3     4      5      6      7      8      9  \
Accident_Severity                                                               
Fatal               104   117    129   115    147    138    121    160    144   
Serious            1139  1290   1606  1599   1757   1740   1765   1626   1791   
Slight             8724  9528  10606  9796  10468  10934  10767  10302  11025   

month                 10     11    12  
Accident_Severity                      
Fatal                151    133    90  
Serious             1762   1631  1071  
Slight             11621  11858  8464

<h2>Is there a correlation between accident severity and number of casualities??</h2>

<h1>Insight No.13 </h1>
<h3>The p-value of zero indicates no statistically significant correlation between the number of casualties and accident severity for slight and serious accidents.</h3>

In [28]:
group1 = accident.loc[accident['Accident_Severity'] == 'Slight', 'Number_of_Casualties']
group2 = accident.loc[accident['Accident_Severity'] == 'Serious', 'Number_of_Casualties']


f_stat, p_value = f_oneway(group1, group2)
print(p_value)


0.0


<h2>What weather conditon have the most number of accident</h2>

<h1>Insight No.14</h1>
<h3>The majority of accidents occurred under 'fine no high winds' weather conditions, accounting for 78.84% of all incidents (520,885 out of 660,679 total accidents), far exceeding other conditions such as 'raining no high winds' (12.06%) and 'snowing no high winds' (0.94%).</h3>

In [29]:
accident['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds           520885
Raining no high winds         79696
Other                         17150
Unknown Weather Condition     14128
Raining + high winds           9615
Fine + high winds              8554
Snowing no high winds          6238
Fog or mist                    3528
Snowing + high winds            885
Name: count, dtype: int64

<h2>What is the average number of accident permonth?</h2>

<h1>Insight No.15</h1>
<h2>The data shows that an average of 55,056 accidents occurred per month.</h2>

In [30]:

monthly_counts = accident['month'].value_counts()


average_accidents_per_month = monthly_counts.mean()

print(f"Average accidents per month: {average_accidents_per_month:.2f}")

Average accidents per month: 55056.58


<h2>What is the average number of accident year? </h2>

<h1>Insight No.16</h1>
<h2>The data shows that an average of 55,056 accidents occurred per year.</h2>

In [31]:

Yearly_counts = accident['Year'].value_counts()


average_accidents_per_Year = Yearly_counts.mean()

print(f"Average accidents per Year: {average_accidents_per_Year:.2f}")

Average accidents per Year: 165169.75


<h2>Which road type has the highest number of accidents?</h2>

<h1>Insight No.17</h1>
<h2>The majority of accidents occurred on single carriageway roads, accounting for 74.49% of all incidents (492,143 out of 660,679 total accidents), far exceeding other road types such as dual carriageways (15.05%) and roundabouts (6.66%).</h2>

In [32]:
accident['Road_Type'].value_counts()

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Unknown Road Type       4520
Name: count, dtype: int64

<h2>What is the distribution of the number of vehicles involved in accidents?</h2>


<h1>Insight No.18</h1>
<h2>Most Accidents Involve Two Vehicles, Accounting for 59.32% of Total Incidents (391,995 out of 660,781)</h2>

In [33]:
accident['Number_of_Vehicles'].value_counts()

Number_of_Vehicles
2     391995
1     200787
3      52806
4      11300
5       2464
6        758
7        296
8        147
9         57
10        27
11        12
14         9
12         7
13         6
16         4
28         1
15         1
32         1
19         1
Name: count, dtype: int64

<h2>Is there a relationship between road type and accident severity?</h2>


<h1>Insight No.19</h1>
<h2>The majority of fatal accidents occurred on single carriageway roads, accounting for 75.36% of all fatalities (6,527 out of 8,661), while slip roads had the lowest number of fatalities at just 0.57% (49 incidents).</h2>

In [34]:
road_severity = accident.groupby(['Road_Type','Accident_Severity']).size()
road_severity.unstack()

Accident_Severity   Fatal  Serious  Slight
Road_Type                                 
Dual carriageway     1815    11746   85863
One way street         95     1655   11809
Roundabout            142     3665   40185
Single carriageway   6527    70059  415557
Slip road              49      611    6381
Unknown Road Type      33      481    4006

<h2>How does number of vehicle involved affect the severity of the accident?</h2>


<h1>Insight No.20</h1>
<h2>single-vehicle accidents are have a higher chance of fatality per incident (1.93%) than two-vehicle ones (0.88%), while four-vehicle accidents top the list at 2.41%, though they happen less often. Using these percentages, you can estimate total fatalities if you know the number of accidents in each group.</h2>

In [35]:
accident.groupby(['Number_of_Vehicles','Accident_Severity']).size().unstack()

Accident_Severity   Fatal  Serious  Slight
Number_of_Vehicles                        
1                    3885    38940  157962
2                    3467    41578  346950
3                     900     5808   46098
4                     272     1340    9688
5                      70      326    2068
6                      28      121     609
7                      17       63     216
8                      12       25     110
9                       1        5      51
10                      2        5      20
11                      2        0      10
12                      0        2       5
13                      0        2       4
14                      3        0       6
15                      0        0       1
16                      1        1       2
19                      0        1       0
28                      1        0       0
32                      0        0       1

<h2>What type of car has the highest fatality rate, and which district has the highest number of fatal accidents?</h2>


<h1>Insight No.21</h1>
<h3>Highland has the highest number of fatal accidents, accounting for 30 incidents (7.1% of all districts). Among these, the majority involve car type, representing 23 out of 30 accidents (76.7%), followed by taxi/private hire cars and light goods vehicles, each accounting for 2 accidents (6.7%). Other vehicle types contribute minimally or not at all to the total.</h3>

In [36]:
si_fatal = during_2020[during_2020['Accident_Severity'] == 'Fatal']
getting_highland = si_fatal[si_fatal['District Area'] == 'Highland']
si_fatal['District Area'].value_counts()


District Area
Highland                    30
Birmingham                  28
Leeds                       24
Aberdeenshire               21
East Riding of Yorkshire    21
                            ..
Harrow                       0
Blaenau Gwent                0
Harlow                       0
Bexley                       0
Hastings                     0
Name: count, Length: 422, dtype: int64

In [37]:
getting_highland['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      23
Taxi/Private hire car                     2
Van / Goods 3.5 tonnes mgw or under       2
Goods 7.5 tonnes mgw and over             1
Motorcycle 50cc and under                 1
Other vehicle                             1
Agricultural vehicle                      0
Bus or coach (17 or more pass seats)      0
Data missing or out of range              0
Goods over 3.5t. and under 7.5t           0
Minibus (8 - 16 passenger seats)          0
Motorcycle 125cc and under                0
Motorcycle over 125cc and up to 500cc     0
Motorcycle over 500cc                     0
Pedal cycle                               0
Ridden horse                              0
Name: count, dtype: int64

<h2>Do night time condition have the highest fatality ? and what district have the highest accident in term of darkness condition</h2>


<h1>Insight No.22</h1>
<h3>Daylight conditions account for the lowest percentage of fatal accidents at 0.9%, while darkness conditions have a higher percentage at 3.1%.</h3>

In [38]:
light_2022 = during_2022[during_2022['Light_Conditions'] != 'Daylight']
darkness = light_2022['Accident_Severity']
darkness.value_counts()

Accident_Severity
Slight     30716
Serious     5350
Fatal        585
Name: count, dtype: int64

In [39]:
light_2022 = during_2022[during_2022['Light_Conditions'] == 'Daylight']
daylight = light_2022['Accident_Severity']
daylight.value_counts()

Accident_Severity
Slight     93377
Serious    13427
Fatal        964
Name: count, dtype: int64